### PR TITLE
Add read-only Engine

### PR DESCRIPTION
**Pull request by s1monw**
This change adds an engine implementation that opens a reader on an
existing index but doesn't permit any refreshes or modifications
to the index.

Relates to #32867
Relates to #32844

### CODE REVIEW

**Comment by s1monw** [comment]
@dnhatn FYI I broken this out of your and my PR since they share pretty much the same functionality into it's own PR

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+    @Override
+    public boolean shouldPeriodicallyFlush() {
+        return false;
+    }
+
+    @Override
+    public SyncedFlushResult syncFlush(String syncId, CommitId expectedCommitId) throws EngineException {
+        // we can't do synced flushes this would require an indexWriter which we don't have
+        return SyncedFlushResult.COMMIT_MISMATCH;
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: hmm... I really wonder if we are better off throwing in an exception here...

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    private final TranslogStats translogStats;
+    private final SearcherManager searcherManager;
+    private final IndexCommit indexCommit;
+    private final Lock indexWriterLock;
+
+    public ReadOnlyEngine(EngineConfig config) {
+        this(config, null, new TranslogStats(0, 0, 0, 0, 0), true);
+    }
+
+    ReadOnlyEngine(EngineConfig config, SeqNoStats seqNoStats, TranslogStats translogStats, boolean obtainLock) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: java docs on obtainLock please.

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.function.BiFunction;
+import java.util.stream.Stream;
+
+/**
+ * A basic read-only engine that allows switching a shard to be true read-only temporarily or permanently.
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: add a note that this be opened concurrently with another engine? It think that's not obvious.

**Review by bleskes — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java (hunk 1 above)
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: yeah lets do that. I think we can cope with it.

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+    @Override
+    public void trimUnreferencedTranslogFiles() { }
+
+    @Override
+    public boolean shouldRollTranslogGeneration() {
+        return false;
+    }
+
+    @Override
+    public void rollTranslogGeneration() throws EngineException { throw new UnsupportedOperationException(); }
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: new lines missing. Also we should assert this is never called. Also no need for `throws EngineException`

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    }
+
+    @Override
+    public boolean isThrottled() {
+        return false;
+    }
+
+    @Override
+    public IndexResult index(Index index) {
+        throw new UnsupportedOperationException();
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: we should assert this is never called (same for the other places here where `UnsupportedOperationException` is thrown), as this indicates a bug.

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+                closedLatch.countDown();
+            }
+        }
+    }
+
+    protected DirectoryReader wrapReader(DirectoryReader reader, ShardId shardId) throws IOException {
+        return ElasticsearchDirectoryReader.wrap(reader, shardId);
+    }
+
+    public static SeqNoStats buildSeqNoStats(SegmentInfos infos) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: why make this public?

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+        return false;
+    }
+
+    @Override
+    public void syncTranslog() {
+    }
+
+    @Override
+    public Closeable acquireRetentionLockForPeerRecovery() {
+        return () -> {};
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: I think this should be an `assert false;` + throw new UnsupportedOperationException

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+        return new Translog.Location(0,0,0);
+    }
+
+    @Override
+    public long getLocalCheckpoint() {
+        return seqNoStats.getLocalCheckpoint();
+    }
+
+    @Override
+    public void waitForOpsToComplete(long seqNo) { }
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: add newline

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+        return 0;
+    }
+
+    @Override
+    public List<Segment> segments(boolean verbose) {
+        return Arrays.asList(getSegmentInfo(lastCommittedSegmentInfos, verbose));
+    }
+
+    @Override
+    public void refresh(String source) throws EngineException {}
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: newline, also no need to write `throws EngineException`

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+
+    @Override
+    public void writeIndexingBuffer() throws EngineException {}
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: newline
also no need to write `throws EngineException`

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+    }
+
+    @Override
+    public SyncedFlushResult syncFlush(String syncId, CommitId expectedCommitId) throws EngineException {
+        // we can't do synced flushes this would require an indexWriter which we don't have
+        return SyncedFlushResult.COMMIT_MISMATCH;
+    }
+
+    @Override
+    public CommitId flush(boolean force, boolean waitIfOngoing) throws EngineException {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: no need for `throws EngineException`

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 12]
```diff
@@ -0,0 +1,349 @@
...
+
+    @Override
+    public void skipTranslogRecovery() {
+    }
+
+    @Override
+    public void trimOperationsFromTranslog(long belowTerm, long aboveSeqNo) throws EngineException { }
+
+    @Override
+    public void maybePruneDeletes() { }
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: newline

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 13]
```diff
@@ -0,0 +1,349 @@
...
+    public Engine recoverFromTranslog(TranslogRecoveryRunner translogRecoveryRunner, long recoverUpToSeqNo) {
+        return this;
+    }
+
+    @Override
+    public void skipTranslogRecovery() {
+    }
+
+    @Override
+    public void trimOperationsFromTranslog(long belowTerm, long aboveSeqNo) throws EngineException { }
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: newline, no need for `throws EngineException`

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java (hunk 7 above)
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: it's totally fine to do a peerRecovery from this  engine.

**Pull request merged**